src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java

```diff
@@ -0,0 +1,122 @@
+    public DuplicateSet(final GATKRead read){
+        reads = new ArrayList<>();
+        init(read);
+        reads.add(read);
+    }
+
+    public void init(GATKRead read){
+        Utils.validate(moleculeId == -1 || moleculeId == getMoleculeID(read),
+                String.format("Inconsisntent molecule IDs: Duplicate set id = %s, read molecule id = %s", moleculeId, getMoleculeID(read)));
+        setMoleduleId(read);umi = new UMI(read);
+        contig = read.getContig();
+        fragmentStart = read.getStart();
+        fragmentEnd = read.getEnd(); // TODO: does this include softclips?
+        paired = false;
+    }
+
+    public List<GATKRead> getReads(){
+        return reads;
+    }
+
+    public boolean sameMolecule(final GATKRead read){
+        return getMoleculeID(read) == moleculeId;
+    }
+
+    public void setMoleduleId(GATKRead read){
+        moleculeId = getMoleculeID(read);
+    }
+
+    public void setMoleduleId(int id){
+        moleculeId = id;
+    }
+
+    /**
+     * Some examples of molecule IDs (MI tag):
+     *
+     * "0/A" (The first molecule in the bam, top (A) strand)
+     * "0/B" (The first molecule in the bam, bottom (B) strand)
+     * "99/A" (100th molecule in the bam, top top (A) strand)
+     *
+     * Top strand is synonymous to "F1R2"
+     * Bottom strand is synonymous to "F2R1"
+     *
+     * Thus only the integer component is relevant for identifying reads that originated from the same molecule.
+     * Should the need arise, we could extend this to distinguish between different strands of the same molecule.
+     */
+    public static int getMoleculeID(final GATKRead read) {
+        final String MITag = read.getAttributeAsString(FGBIO_MOLECULAR_IDENTIFIER_TAG);
+        return Integer.parseInt(MITag.split(FGBIO_MI_TAG_DELIMITER)[0]);
+    }
+
+    /** Returns true if the read was properly added to the duplicate set **/
+    public boolean addRead(final GATKRead read){
+        if (reads.isEmpty()){
+            init(read);
+            reads.add(read);
+            return true;
+        }
+
+        if (sameMolecule(read)){
+            reads.add(read);
+            if (read.getStart() < fragmentStart){
+                fragmentStart = read.getStart();
+            }
+            if (read.getEnd() > fragmentEnd){
+                fragmentEnd = read.getEnd();
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public int getFragmentStart(){
+        return fragmentStart;
+    }
+
+    public int getFragmentEnd(){
+        return fragmentEnd;
+    }
+
+    public int getMoleculeId() { return moleculeId; }
+
+    public SimpleInterval getDuplicateSetInterval(){
+        return new SimpleInterval(contig, fragmentStart, fragmentEnd);
+    }
+
+    public boolean hasValidInterval(){
+        return SimpleInterval.isValid(contig, fragmentStart, fragmentEnd);
+    }
+
+    public static List<String> getMolecularIDs(final List<GATKRead> reads) {
```

Comment:
Ok, I understand this now... you need clear javadoc and probably to rename this method since it's very strange that getMolecularID on 2 reads can produce the same result but getMolecularIDs then has 2 values.